oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java

```diff
@@ -1246,6 +1349,256 @@ private void collectRevisionsOlderThan24hAndBetweenCheckpoints(final NodeDocumen
+                final DocumentNodeState traversedNode = bundledNodeStates.get(Path.fromString(unbundledPath));
+                if (traversedNode == null) {
+                    log.error("collectUnusedPropertyRevisions : could not find traversed node for bundled key {} unbundledPath {} in doc {}",
+                            propName, unbundledPath, doc.getId());
+                    continue;
+                }
+                final PropertyState traversedProperty = traversedNode.getProperty(unbundledPropName);
+                if (traversedProperty == null) {
+                    log.error("collectUnusedPropertyRevisions : could not get property {} from traversed node {}",
+                            unbundledPropName, traversedNode.getPath());
+                    continue;
+                }
+                deletedRevsCount += removeUnusedPropertyEntries(doc, traversedNode, updateOp, escapedPropName,
+                        (r) -> { updateOp.removeMapEntry(escapedPropName, r); return null;},
+                        allKeepRevs);
+            }
+
+            // phase 3 : "_deleted"
+            int numDeleted = removeUnusedPropertyEntries(doc, traversedMainNode,
+                    updateOp, NodeDocument.DELETED,
+                    (r) -> { NodeDocument.removeDeleted(updateOp, r); return null;} ,
+                    allKeepRevs);
+            deletedRevsCount += numDeleted;
+            return deletedRevsCount;
+        }
+
+        private int collectUnusedInternalPropertyRevisions(final NodeDocument doc,
+                final GCPhases phases, final UpdateOp updateOp,
+                final DocumentNodeState traversedMainNode,
+                final Set<Revision> toKeepUserPropRevs,
+                int deletedRevsCount) {
+            boolean hasUnmergedBranchCommits = false;
+            for (Revision localBc : doc.getLocalBranchCommits()) {
+                if (!isCommitted(nodeStore.getCommitValue(localBc, doc))) {
+                    hasUnmergedBranchCommits = true;
+                }
+            }
+            if (deletedRevsCount == 0 && !hasUnmergedBranchCommits) {
+                return deletedRevsCount;
+            }
+            // if we did some rev deletion OR there are unmerged BCs, then let's deep-dive
+            Set<Revision> allRequiredRevs = new HashSet<>(toKeepUserPropRevs);
+            for (Revision revision : doc.getLocalMap(NodeDocument.COLLISIONS).keySet()) {
+                if (!allRequiredRevs.contains(revision)) {
+                    Operation has = updateOp.getChanges().get(new Key(NodeDocument.COLLISIONS, revision));
+                    if (has != null) {
+                        // then skip
+                        continue;
+                    }
+                    NodeDocument.removeCollision(updateOp, revision);
+                    deletedRevsCount++;
+                }
+            }
```

Comment:
```suggestion
            // "_collisions"
            deletedRevsCount += getDeletedRevsCount(doc.getLocalCommitRoot().keySet(), updateOp, allRequiredRevs, COLLISIONS, NodeDocument::removeCollision);
```

Reply:
incorporated - plus fixed `getLocalCommitRoot()` to `getLocalMap(..)` in https://github.com/apache/jackrabbit-oak/pull/1372/commits/4bbb8ed82c9ac7ae53994f9de1cbed15ca515718